paddle/fluid/imperative/layer.cc

```diff
@@ -240,189 +350,452 @@ std::unique_ptr<VarBase> VarBase::NewVarBase(const platform::Place& dst_place,
+      }
+    }
   }
 }
 
-void OpBase::RegisterBackwardHooks(const py::object& callable) {
-  VLOG(3) << "Register backward hooks " << trace_id_;
+void AutoGradImpl::PrepareDeps() {
+  PADDLE_ENFORCE(op_deps_.empty(), "Op deps must be initialized here");
+  PADDLE_ENFORCE(accumulators_.empty(),
+                 "Accumulators must be initialized here");
+
+  std::queue<OpBase*> q;
+  std::unordered_set<OpBase*> visited;
+  q.push(op_);
+  visited.insert(op_);
+
+  while (!q.empty()) {
+    auto* cur_op = q.front();
+    q.pop();
+    VLOG(2) << "Checking grads of op " << cur_op->Type();
+
+    if (CheckBackwardInputs(cur_op)) {
+      // TODO(zjl): clear ops that do not need grad before running autograd
+      VLOG(2) << "Stop checking preceding ops of " << cur_op->Type()
+              << " because all of its backward inputs is stop_gradient=True";
+      continue;
+    }
 
-  // TODO(minqiyang): check the callable format
-  backward_hooks_.push_back(callable);
+    CheckBackwardOutputs(cur_op);
+
+    auto& preceding_ops = cur_op->PrecedingOps();
+    for (auto* preceding_op : preceding_ops) {
+      PADDLE_ENFORCE_NOT_NULL(preceding_op);
+      ++op_deps_[preceding_op];
+      if (visited.count(preceding_op) == 0) {
+        visited.insert(preceding_op);
+        q.push(preceding_op);
+      }
+    }
+  }
 }
 
-void VarBase::RunBackward(const detail::BackwardStrategy& bck_stratedy) {
-  if (!pre_op_) return;
-  platform::RecordEvent record_event("Imperative Backward");
-  VLOG(3) << "start backward";
-  auto grads_t = grads_->var_->GetMutable<framework::LoDTensor>();
-  operators::math::set_constant(
-      *(platform::DeviceContextPool::Instance().Get(
-          var_->GetMutable<framework::LoDTensor>()->place())),
-      grads_t, 1.0);
-
-  PADDLE_ENFORCE(
-      grads_ ==
-      pre_op_->output_vars_[pre_op_out_name_][pre_op_out_idx_]->grads_);
-  Autograd().RunBackward(this, bck_stratedy);
+void AutoGradImpl::operator()() {
+  std::queue<OpBase*> q;
+  q.push(op_);
+
+  while (!q.empty()) {
+    OpBase* cur_op = q.front();
+    q.pop();
+
+    // Step 1: Run Backward
+    auto& grad_descs = cur_op->GradOpDescs();
+    auto& bwd_ins = cur_op->BackwardInputs();
+    auto& bwd_outs = cur_op->BackwardOutputs();
+    size_t grad_op_num = grad_descs.size();
+
+    PADDLE_ENFORCE_EQ(grad_op_num, bwd_ins.size());
+    PADDLE_ENFORCE_EQ(grad_op_num, bwd_outs.size());
+
+    for (size_t i = 0; i < grad_op_num; ++i) {
+      NameVarBaseMap tmp_outs;
+      // A var may be coresponding to several grad var in one op
+      std::unordered_map<VarBase*, std::vector<std::shared_ptr<VarBase>>>
+          var_map;
+      size_t counter = 0;
+      for (auto& bwd_out : bwd_outs[i]) {
+        auto& tmp_var_list = tmp_outs[bwd_out.first];
+        tmp_var_list.reserve(bwd_out.second.size());
+        for (auto& var : bwd_out.second) {
+          auto tmp_var = std::make_shared<VarBase>(false);  // Do not need grad
```

Comment:
why share_ptr here?

Reply:
Because `Tracer` needs the input type to be `std::shared_ptr`.

Reply:
Because `Tracer` needs the input type to be `std::shared_ptr`.